pallets/withdraw-teleport/src/lib.rs

```diff
@@ -185,23 +186,29 @@ impl<T: Config> Pallet<T> {
 
 		//Build the message to execute on origin.
 		let assets: MultiAssets = assets.into();
-		let message: Xcm<<T as frame_system::Config>::RuntimeCall> =
-			Xcm(vec![WithdrawAsset(assets.clone()), BurnAsset(assets)]);
+		let message: Xcm<<T as frame_system::Config>::RuntimeCall> = Xcm(vec![
+			// Burn the native asset.
+			WithdrawAsset(assets.clone()),
+			BuyExecution { fees: native_asset, weight_limit: Unlimited },
+			BurnAsset(assets),
+			// Burn the fee asset derivative.
+			WithdrawAsset(fee_asset.clone()),
+			BurnAsset(fee_asset.clone()),
+		]);
 
 		// Build the message to send to be executed.
 		// Set WeightLimit
 		// TODO: Implement weight_limit calculation with final instructions.
 		let weight_limit: WeightLimit = Unlimited;
 		let xcm_to_send: Xcm<()> = Xcm(vec![
-			// There are currently no limitations on the amount of fee assets to withdraw.
-			// Since funds are withdrawn from the Sovereign Account of the origin, chains must be
-			// aware of this and implement a mechanism to prevent draining.
-			WithdrawAsset(fee_asset),
+			// User must have the derivative of fee_asset on origin.
+			WithdrawAsset(fee_asset.clone()),
 			BuyExecution { fees, weight_limit },
-			ReceiveTeleportedAsset(foreign_assets.clone()),
-			// Intentionally trap ROC to avoid exploit of draining Sovereing Account
-			// by depositing withdrawn ROC on beneficiary.
-			DepositAsset { assets: MultiAssetFilter::Definite(foreign_assets), beneficiary },
+			ReceiveTeleportedAsset(foreing_assets.clone()),
+			// We can deposit funds since they were both withdrawn on origin.
+			DepositAsset { assets: MultiAssetFilter::Definite(foreing_assets), beneficiary },
+			RefundSurplus,
+			DepositAsset { assets: MultiAssetFilter::Definite(fee_asset), beneficiary },
```

Comment:
But you don't know how much of `fee_asset` remains, since part of it was moved into fee payment. I think you need an `AllOf` variant.

Reply:
Ah, I had the misconception that this `MultiAssetFilter` deposited "whatever you have left of x asset" but it takes also in consideration the amount? Will fix.